<launch>

<arg name="loop_rate" default="30" />
<arg name="navion_version" default="1_1" />

<!-- Camera -->
  <node pkg="usb_cam" type="usb_cam_node" name="usb_cam" output="screen" >
          <param name="video_device" value="/dev/video0" />
          <param name="image_width" value="400" />
          <param name="image_height" value="380" />
          <param name="pixel_format" value="yuyv" />
          <param name="camera_frame_id" value="usb_cam" />
          <param name="io_method" value="mmap"/>
          <param name="autofocus" value="false"/>
          <param name="camera_info_url" value="file:///$(find endoscopy_demo)/config/cameras/endoscope_v1001_cam_calib.yaml"/>
  </node>

  <group ns="usb_cam">
        <node pkg="image_proc" type="image_proc" name="image_proc" />
  </group>


<!-- Optical flow  -->
  <node name="flow" pkg="opencv_apps" type="phase_corr" >
      <remap from="image" to="/usb_cam/image_raw" />
      <param name="use_camera_info" value="false" />
      <param name="debug_view" value="false" />
      <param name="block_size" value="10" />

  </node>
    
<!-- Navigation node  -->
  <node name="pov_nav" pkg="pov_mag_navigation" type="pov_navigation_node.py" output="screen"  >
        <remap from="/desired_field" to="/backward_model/field"/>
        <param name="~rate" value="$(arg loop_rate)"/>
        <rosparam command="load" file="$(find pov_mag_navigation)/param/parameters_pov_transnasal.yaml" />
  </node>

<!-- PS navi -->
  <node respawn="true" pkg="joy" type="joy_node" name="joy_node" >
      <param name="autorepeat_rate" value="$(arg loop_rate)"/>
   </node>
  <node name="ps_navi_controller" pkg="pov_mag_navigation" type="ps_navi_controller" output="screen" >
  </node>

<!-- Joystick axes feedback  -->
  <node name="feedback" pkg="pov_mag_navigation" type="feedback_plot.py" output="screen"  >
  </node>

<!-- Dynamic reconfigure -->
  <node pkg="rqt_reconfigure" type="rqt_reconfigure" name="rqt_reconfigure" >
  </node>

<!-- Rviz -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find pov_mag_navigation)/rviz/pov_nav_navion.rviz" />

<!-- eMNS -->
  <arg name="cal_path" default="$(find mpem)/cal/Navion_2_Calibration_24-02-2020.yaml" />
  <arg name="emns_param_path" default="$(find mag_manip)/params/Navion_2.yaml" />

  <node name="mns_broadcaster" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 1 world mns 100" respawn="true" />

  <node name="nav_controller" pkg="nav_controller" type="navion_controller_node" output="screen" respawn="true">
          <rosparam command="load" file="$(find pov_mag_navigation)/launch/parameters_navion.yaml" />
          <rosparam command="load" file="$(find nav_controller)/param/navion_$(arg navion_version).yaml" />
  </node>

  <node name="backward_model" pkg="mag_calculator" type="backward_model" output="screen" respawn="true">
        <param name="calibration_path" type="string" value="$(arg cal_path)" />
        <param name="is_field_inverted" type="bool" value="false" />
        <param name="max_field_intensity" type="double" value="0.025" />
        <remap from="/backward_model/currents" to="navion/target_currents"/>
  </node>


</launch>